lib/common/init.js

```diff
@@ -8,26 +9,31 @@ process.atomBinding = require('./atom-binding-setup')(process.binding, process.t
 // which would delay the callbacks for arbitrary long time. So we should
 // initiatively activate the uv loop once setImmediate and process.nextTick is
 // called.
-var wrapWithActivateUvLoop = function (func) {
-  return function () {
-    process.activateUvLoop()
-    return func.apply(this, arguments)
+const asyncHook = asyncHooks.createHook({
+  init (asyncId, type, triggerAsyncId) {
+    switch (type.toUpperCase()) {
+      case 'IMMEDIATE':
+      case 'TIMEOUT':
+      case 'TICKOBJECT':
+        process.activateUvLoop()
+        break
+      default:
+        break
+    }
   }
-}
-
-process.nextTick = wrapWithActivateUvLoop(process.nextTick)
-
-global.setImmediate = wrapWithActivateUvLoop(timers.setImmediate)
+})
+asyncHook.enable()
 
+global.setImmediate = timers.setImmediate
 global.clearImmediate = timers.clearImmediate
 
 if (process.type === 'browser') {
   // setTimeout needs to update the polling timeout of the event loop, when
   // called under Chromium's event loop the node's event loop won't get a chance
   // to update the timeout, so we have to force the node's event loop to
   // recalculate the timeout in browser process.
-  global.setTimeout = wrapWithActivateUvLoop(timers.setTimeout)
-  global.setInterval = wrapWithActivateUvLoop(timers.setInterval)
+  global.setTimeout = timers.setTimeout
+  global.setInterval = timers.setInterval
```

Comment:
We can remove the code in this `if` condition now, Node's bootstrap code already set the globals. We should only need:

```js
if (process.type === 'renderer') {
  global.setImmediate = timers.setImmediate
  global.clearImmediate = timers.clearImmediate
}
```

Reply:
@zcbenz done (assuming this is what you meant)